File: graphs & tries/207.py
# Google, Amazon, Microsoft, Meta, Apple
# DFS
# Directed Graph
# If cycle return False else True
# Tc = O(V+E)
# Sc = O(V+E)
from collections import defaultdict
class Solution:
    def isCycleDFS(self, adj, u, visited, inRecursion):
        visited[u] = True
        inRecursion[u] = True
        for v in adj[u]:
            if not visited[v] and self.isCycleDFS(adj, v, visited, inRecursion):
                return True
            elif inRecursion[v]:
                return True
        inRecursion[u] = False
        return False

    def canFinish(self, numCourses, prerequisites):
        adj = defaultdict(list)
        visited = [False] * numCourses
        inRecursion = [False] * numCourses
        for a, b in prerequisites:
            adj[b].append(a)
        for i in range(numCourses):
            if not visited[i] and self.isCycleDFS(adj, i, visited, inRecursion):
                return False
        return True

File: graphs & tries/207.js
class Solution{
    isCycleDFS(adj, u, visited, inRecursion){
        visited[u] = true
        inRecursion[u] = true
        for(let v of adj[u]){
            if(!visited[v] && this.isCycleDFS(adj, v, visited, inRecursion)){
                return true
            }
            else if(inRecursion[v]){
                return true
            }
        }
        inRecursion[u] = false
        return false
    }
    canFinish(numCourses, prerequisites){
        const adj = Array.from({length: numCourses}, () =>[])
        const visited = new Array(numCourses).fill(false)
        const inRecursion = new Array(numCourses).fill(false)
        for(let [a,b] of prerequisites){
            adj[b].push(a)
        }
        for(let i = 0; i <numCourses; i++){
            if(!visited[i] && this.isCycleDFS(adj, i, visited, inRecursion)){
                return false
            }
        }
        return true
    }
}